File: src/intcode.rs
use std::ops::{Index, IndexMut};

#[derive(Debug, Clone, Copy)]
enum InstructionArgument {
    Position(usize),
    Immediate(i64),
    Relative(isize),
}

impl InstructionArgument {
    fn from(mode: u8, value: i64) -> Self {
        match mode {
            0 => InstructionArgument::Position(value as usize),
            1 => InstructionArgument::Immediate(value),
            2 => InstructionArgument::Relative(value as isize),
            _ => panic!("unsupported instruction argument value {}", mode),
        }
    }
}

#[derive(Debug)]
struct Memory {
    ram: Vec<i64>,
    program_size: usize,
    program_counter: usize,
    relative_base: usize,
}

impl Memory {
    fn load_program(program: &[i64]) -> Self {
        let mut ram = vec![0; program.len() * 2];
        ram[..program.len()].clone_from_slice(&program[..]);
        Memory {
            ram,
            program_size: program.len(),
            program_counter: 0,
            relative_base: 0,
        }
    }

    fn get_instruction(&self) -> Option<i64> {
        self.ram.get(self.program_counter).cloned()
    }

    fn get(&self, arg: InstructionArgument) -> i64 {
        match arg {
            InstructionArgument::Position(addr) => self.ram[addr],
            InstructionArgument::Immediate(value) => value,
            InstructionArgument::Relative(rel_addr) => {
                let addr = ((self.relative_base as isize) + rel_addr) as usize;
                self.ram[addr]
            }
        }
    }

    fn set(&mut self, arg: InstructionArgument, value: i64) {
        match arg {
            InstructionArgument::Position(addr) => self.ram[addr] = value,
            InstructionArgument::Immediate(_) => {
                panic!("writes with immediate arguments are not supported")
            }
            InstructionArgument::Relative(rel_addr) => {
                let addr = ((self.relative_base as isize) + rel_addr) as usize;
                self.ram[addr] = value;
            }
        };
    }

    fn checked_adjust_relative_base(&mut self, modifier: i64) {
        if modifier < 0 {
            self.relative_base -= i64::abs(modifier) as usize;
        } else {
            self.relative_base += modifier as usize;
        }
    }

    fn dump(&self) -> String {
        format!("{:?}", self.ram)
    }
}

impl Index<usize> for Memory {
    type Output = i64;

    fn index(&self, index: usize) -> &Self::Output {
        &self.ram[index]
    }
}

impl IndexMut<usize> for Memory {
    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
        &mut self.ram[index]
    }
}

#[derive(Debug)]
enum ProgramCounter {
    Stride(usize),
    Interrupt(Interrupt),
    Jump(usize),
    Halt,
}

#[derive(Debug)]
pub enum Interrupt {
    Halted,
    WaitingForInput,
    Output(i64),
}

enum Opcode {
    Add = 1,
    Multiply = 2,
    Load = 3,
    Output = 4,
    JumpIfTrue = 5,
    JumpIfFalse = 6,
    LessThan = 7,
    Equals = 8,
    SetRelativeBase = 9,
    Halt = 99,
}

impl From<u8> for Opcode {
    fn from(byte: u8) -> Self {
        match byte {
            1 => Opcode::Add,
            2 => Opcode::Multiply,
            3 => Opcode::Load,
            4 => Opcode::Output,
            5 => Opcode::JumpIfTrue,
            6 => Opcode::JumpIfFalse,
            7 => Opcode::LessThan,
            8 => Opcode::Equals,
            9 => Opcode::SetRelativeBase,
            99 => Opcode::Halt,
            _ => panic!("{} is not a valid opcode", byte),
        }
    }
}

#[derive(Debug)]
struct Input {
    arg: InstructionArgument,
    value: Option<i64>,
}

#[derive(Debug)]
pub struct Intcode {
    memory: Memory,
    pending_input: Option<Input>,
    pending_output: Option<i64>,
}

impl Intcode {
    pub fn new(program: &[i64]) -> Self {
        Intcode {
            memory: Memory::load_program(program),
            pending_input: None,
            pending_output: None,
        }
    }

    pub fn dump_memory(&self) -> String {
        self.memory.dump()
    }

    pub fn run(&mut self) -> Interrupt {
        loop {
            let instruction = self.memory.get_instruction();
            let call = match instruction {
                Some(code) => self.call(code),
                None => panic!("Address {} does not exist", self.memory.program_counter),
            };
            match call {
                ProgramCounter::Stride(stride) => self.memory.program_counter += stride,
                ProgramCounter::Interrupt(cause) => return cause,
                ProgramCounter::Jump(addr) => self.memory.program_counter = addr,
                ProgramCounter::Halt => return Interrupt::Halted,
            };
        }
    }

    pub fn set_input(&mut self, input: i64) {
        if let Some(pending_input) = self.pending_input.take() {
            self.pending_input = Some(Input {
                arg: pending_input.arg,
                value: Some(input),
            });
        }
    }

    fn get_arg(&self, code: i64, index: usize) -> InstructionArgument {
        let power = (index as u32) + 2;
        let mask = i64::pow(10, power);
        let mode = ((code / mask) % 10) as u8;
        let value = self.memory[self.memory.program_counter + index + 1];
        InstructionArgument::from(mode, value)
    }

    fn call(&mut self, code: i64) -> ProgramCounter {
        let get_two = || (self.get_arg(code, 0), self.get_arg(code, 1));
        let get_three = || {
            (
                self.get_arg(code, 0),
                self.get_arg(code, 1),
                self.get_arg(code, 2),
            )
        };
        let opcode: Opcode = From::from((code % 100) as u8);
        // let args = mem.get_args(code);
        match opcode {
            Opcode::Add => {
                let (augend, addend, sum) = get_three();
                self.add(augend, addend, sum);
                ProgramCounter::Stride(4)
            }
            Opcode::Multiply => {
                let (multiplier, multiplicand, product) = get_three();
                self.mult(multiplier, multiplicand, product);
                ProgramCounter::Stride(4)
            }
            Opcode::Load => match self.pending_input {
                Some(Input {
                    arg,
                    value: Some(input),
                }) => {
                    self.ld(arg, input);
                    self.pending_input = None;
                    ProgramCounter::Stride(2)
                }
                _ => {
                    let arg = self.get_arg(code, 0);
                    self.pending_input = Some(Input { arg, value: None });
                    ProgramCounter::Interrupt(Interrupt::WaitingForInput)
                }
            },
            Opcode::Output => match self.pending_output {
                Some(_) => {
                    self.pending_output = None;
                    ProgramCounter::Stride(2)
                }
                None => {
                    let arg = self.get_arg(code, 0);
                    let output = self.output_arg(arg);
                    self.pending_output = Some(output);
                    ProgramCounter::Interrupt(Interrupt::Output(output))
                }
            },
            Opcode::JumpIfTrue => {
                let (arg0, arg1) = get_two();
                self.jump_if_true(arg0, arg1)
            }
            Opcode::JumpIfFalse => {
                let (arg0, arg1) = get_two();
                self.jump_if_false(arg0, arg1)
            }
            Opcode::LessThan => {
                let (lhs, rhs, addr) = get_three();
                self.less_than(lhs, rhs, addr);
                ProgramCounter::Stride(4)
            }
            Opcode::Equals => {
                let (lhs, rhs, addr) = get_three();
                self.equals(lhs, rhs, addr);
                ProgramCounter::Stride(4)
            }
            Opcode::SetRelativeBase => {
                let arg = self.get_arg(code, 0);
                self.memory
                    .checked_adjust_relative_base(self.memory.get(arg));
                ProgramCounter::Stride(2)
            }
            Opcode::Halt => ProgramCounter::Halt,
        }
    }

    fn add(
        &mut self,
        augend: InstructionArgument,
        addend: InstructionArgument,
        sum: InstructionArgument,
    ) {
        self.memory
            .set(sum, self.memory.get(augend) + self.memory.get(addend));
    }

    fn mult(
        &mut self,
        multiplier: InstructionArgument,
        multiplicand: InstructionArgument,
        product: InstructionArgument,
    ) {
        self.memory.set(
            product,
            self.memory.get(multiplier) * self.memory.get(multiplicand),
        );
    }

    fn ld(&mut self, addr: InstructionArgument, input: i64) {
        self.memory.set(addr, input);
    }

    fn output_arg(&mut self, addr: InstructionArgument) -> i64 {
        self.memory.get(addr)
    }

    fn jump_if_true(
        &mut self,
        arg0: InstructionArgument,
        arg1: InstructionArgument,
    ) -> ProgramCounter {
        if self.memory.get(arg0) != 0 {
            ProgramCounter::Jump(self.memory.get(arg1) as usize)
        } else {
            ProgramCounter::Stride(3)
        }
    }

    fn jump_if_false(
        &mut self,
        arg0: InstructionArgument,
        arg1: InstructionArgument,
    ) -> ProgramCounter {
        if self.memory.get(arg0) == 0 {
            ProgramCounter::Jump(self.memory.get(arg1) as usize)
        } else {
            ProgramCounter::Stride(3)
        }
    }

    fn less_than(
        &mut self,
        lhs: InstructionArgument,
        rhs: InstructionArgument,
        addr: InstructionArgument,
    ) {
        if self.memory.get(lhs) < self.memory.get(rhs) {
            self.memory.set(addr, 1);
        } else {
            self.memory.set(addr, 0);
        }
    }

    fn equals(
        &mut self,
        lhs: InstructionArgument,
        rhs: InstructionArgument,
        addr: InstructionArgument,
    ) {
        if self.memory.get(lhs) == self.memory.get(rhs) {
            self.memory.set(addr, 1);
        } else {
            self.memory.set(addr, 0);
        }
    }
}

File: Sources/2019/Utilities/Intcode.swift
//
//  Intcode.swift
//  
//
//  Created by Aaron Sky on 11/18/21.
//

import Foundation

struct Intcode: CustomDebugStringConvertible {
    private var memory: Memory
    private var pendingInput: Intcode.Input?
    private var pendingOutput: Int?

    var debugDescription: String {
        memory.debugDescription
    }

    init(program: [Int]) {
        memory = Memory(program: program)
    }

    @discardableResult
    mutating func run() -> Interrupt {
        while true {
            let call: ProgramCounter
            if let instruction = memory.instruction {
                do {
                    call = try self.call(instruction)
                } catch {
                    fatalError(error.localizedDescription)
                }
            } else {
                fatalError("Address \(memory.programCounter) does not exist")
            }
            switch call {
            case .stride(let stride):
                memory.programCounter += stride
            case .interrupt(let cause):
                return cause
            case .jump(let addr):
                memory.programCounter = addr
            case .halt:
                return .halted
            }
        }
    }

    mutating func call(_ code: Int) throws -> ProgramCounter {
        guard let opcode = Opcode(rawValue: code % 100) else {
            throw Opcode.Error.invalid(code)
        }

        switch opcode {
        case .add:
            let (augend, addend, sum) = (
                self.getArgument(code: code, index: 0),
                self.getArgument(code: code, index: 1),
                self.getArgument(code: code, index: 2)
        )
            add(augend, addend, sum)
            return .stride(4)
        case .multiply:
            let (multiplier, multiplicand, product) = (
                    self.getArgument(code: code, index: 0),
                    self.getArgument(code: code, index: 1),
                    self.getArgument(code: code, index: 2)
            )
            mult(multiplier, multiplicand, product)
            return .stride(4)
        case .load:
            if let pendingInput = pendingInput, let input = pendingInput.value {
                ld(pendingInput.arg, input)
                self.pendingInput = nil
                return .stride(2)
            } else {
                let arg = getArgument(code: code, index: 0)
                self.pendingInput = Input(arg: arg, value: nil)
                return .interrupt(.waitingForInput)
            }
        case .output:
            if pendingOutput != nil {
                self.pendingOutput = nil
                return .stride(2)
            } else {
                let arg = getArgument(code: code, index: 0)
                let output = outputArgument(arg)
                self.pendingOutput = output
                return .interrupt(.output(output))
            }
        case .jumpIfTrue:
            let (arg0, arg1) = (
                self.getArgument(code: code, index: 0),
                self.getArgument(code: code, index: 1)
            )
            return jumpIfTrue(arg0, arg1)
        case .jumpIfFalse:
            let (arg0, arg1) = (
                self.getArgument(code: code, index: 0),
                self.getArgument(code: code, index: 1)
            )
            return jumpIfFalse(arg0, arg1)
        case .lessThan:
            let (lhs, rhs, addr) = (
                self.getArgument(code: code, index: 0),
                self.getArgument(code: code, index: 1),
                self.getArgument(code: code, index: 2)
        )
            lessThan(lhs, rhs, addr)
            return .stride(4)
        case .equals:
            let (lhs, rhs, addr) = (
                self.getArgument(code: code, index: 0),
                self.getArgument(code: code, index: 1),
                self.getArgument(code: code, index: 2)
        )
            equals(lhs, rhs, addr)
            return .stride(4)
        case .setRelativeBase:
            let arg = getArgument(code: code, index: 0)
            memory.adjustRelativeBase(memory[arg])
            return .stride(2)
        case .halt:
            return .halt
        }
    }

    func getArgument(code: Int, index: Int) -> InstructionArgument {
        let power = index + 2
        let mask = pow(10, Double(power))
        let mode = (code / Int(mask)) % 10
        let value = memory[memory.programCounter + index + 1]

        return InstructionArgument(mode: mode, value: value)
    }

    mutating func set(input value: Int) {
        guard let input = pendingInput else {
            return
        }

        pendingInput = Input(arg: input.arg, value: value)
    }

    mutating func set(address: Int, value: Int) {
        memory[address] = value
    }

    mutating func add(_ augend: InstructionArgument, _ addend: InstructionArgument, _ sum: InstructionArgument) {
        memory[sum] = memory[augend] + memory[addend]
    }

    mutating func mult(_ multiplier: InstructionArgument, _ multiplicand: InstructionArgument, _ product: InstructionArgument) {
        memory[product] = memory[multiplier] * memory[multiplicand]
    }

    mutating func ld(_ addr: InstructionArgument, _ input: Int) {
        memory[addr] = input
    }

    mutating func outputArgument(_ addr: InstructionArgument) -> Int {
        memory[addr]
    }

    mutating func jumpIfTrue(_ arg0: InstructionArgument, _ arg1: InstructionArgument) -> ProgramCounter {
        if memory[arg0] != 0 {
            return .jump(memory[arg1])
        } else {
            return .stride(3)
        }
    }

    mutating func jumpIfFalse(_ arg0: InstructionArgument, _ arg1: InstructionArgument) -> ProgramCounter {
        if memory[arg0] == 0 {
            return .jump(memory[arg1])
        } else {
            return .stride(3)
        }
    }

    mutating func lessThan(_ lhs: InstructionArgument, _ rhs: InstructionArgument, _ addr: InstructionArgument) {
        if memory[lhs] < memory[rhs] {
            memory[addr] = 1
        } else {
            memory[addr] = 0
        }
    }

    mutating func equals(_ lhs: InstructionArgument, _ rhs: InstructionArgument, _ addr: InstructionArgument) {
        if memory[lhs] == memory[rhs] {
            memory[addr] = 1
        } else {
            memory[addr] = 0
        }
    }

    struct Memory: CustomDebugStringConvertible {
        var ram: [Int]
        var programSize: Int
        var programCounter: Int = 0
        var relativeBase: Int = 0

        var instruction: Int? {
            ram.indices.contains(programCounter) ? ram[programCounter] : nil
        }

        var debugDescription: String {
            "\(ram)"
        }

        init(program: [Int]) {
            ram = .init(repeating: 0, count: program.count * 2)
            ram.replaceSubrange(..<program.count, with: program)
            programSize = program.count
        }

        subscript(arg: InstructionArgument) -> Int {
            get {
                switch arg {
                case .position(let addr):
                    return ram[addr]
                case .immediate(let value):
                    return value
                case .relative(let addr):
                    return ram[relativeBase + addr]
                }
            }
            set {
                switch arg {
                case .position(let addr):
                    ram[addr] = newValue
                case .immediate:
                    fatalError("writes with immediate arguments are not supported")
                case .relative(let addr):
                    ram[relativeBase + addr] = newValue
                }
            }
        }

        subscript(_ index: Array<Int>.Index) -> Int {
            get {
                ram[index]
            }
            set {
                ram[index] = newValue
            }
        }

        mutating func adjustRelativeBase(_ modifier: Int) {
            if modifier < 0 {
                relativeBase -= abs(modifier)
            } else {
                relativeBase += modifier
            }
        }
    }

    struct Input {
        var arg: InstructionArgument
        var value: Int?
    }

    enum InstructionArgument {
        case position(Int)
        case immediate(Int)
        case relative(Int)

        init(mode: Int, value: Int) {
            switch mode {
            case 0:
                self = .position(value)
            case 1:
                self = .immediate(value)
            case 2:
                self = .relative(value)
            default:
                fatalError("unsupported instruction argument value \(mode)")
            }
        }
    }

    enum ProgramCounter {
        case stride(Int)
        case interrupt(Interrupt)
        case jump(Int)
        case halt
    }

    enum Interrupt {
        case halted
        case waitingForInput
        case output(Int)
    }

    enum Opcode: Int {
        case add = 1
        case multiply = 2
        case load = 3
        case output = 4
        case jumpIfTrue = 5
        case jumpIfFalse = 6
        case lessThan = 7
        case equals = 8
        case setRelativeBase = 9
        case halt = 99

        enum Error: Swift.Error {
            case invalid(Int)
        }
    }
}
